### PR TITLE
Innerlink functionality 

### PR DESCRIPTION
**Pull request by rohanrjohn**

### CODE REVIEW

File: src/LinkSetURLCommand.js [hunk 1]
```diff
@@ -43,15 +43,45 @@ class LinkSetURLCommand extends UICommand {
...
+  waitForUserInput = async (
     state: EditorState,
     dispatch: ?(tr: Transform) => void,
     view: ?EditorView,
     event: ?SyntheticEvent<>
   ): Promise<any> => {
-    if (this._popUp) {
-      return Promise.resolve(undefined);
-    }
+    // if (this._popUp) {
```
Comment (Tezra) [inline src/LinkSetURLCommand.js]: deleted commented code

File: src/ui/LinkURLEditor.js [hunk 2]
```diff
@@ -61,33 +88,123 @@ class LinkURLEditor extends React.PureComponent<any, any> {
...
+    const TocNode = view.state.doc.nodeAt(tocNodePosition_);
+    if (TocNode && (TocNode.attrs.innerLink == null || TocNode.attrs.innerLink == "" || TocNode.attrs.innerLink == undefined)) {
+
+      const nodeUUID = uuid();
+      const texthash = '#';
+      let nodeAttrs = TocNode.attrs;
+      const nodeconcat_UUID = texthash.concat(nodeUUID);
+      nodeAttrs.innerLink = nodeconcat_UUID;
+      tr.setNodeMarkup(tocNodePosition_, undefined, nodeAttrs);
+      var textContent = nodeconcat_UUID.concat('INNER______LINK', textContent_);
```
Comment (Tezra) [inline src/ui/LinkURLEditor.js]: 'INNER______LINK' should probably be a constant.

File: src/ui/czi-link-tooltip.css [hunk 3]
```diff
@@ -69,3 +69,57 @@
     display: none;
   }
 }
+
+/* .czi-link-tooltip-toc {
```
Comment (Tezra) [inline src/ui/czi-link-tooltip.css]: remove unused block

**Review by Tezra — CHANGES_REQUESTED** [review]
Shouldn't everything related to "StyleRuntime" be part of the Style plugin?

**Comment by rohanrjohn** [comment]
> Shouldn't everything related to "StyleRuntime" be part of the Style plugin?

InnerLink is currently applied to toc items which are distinguished by stylename property. Currently in Licit, we are only passing CustomLicitRuntime and the StyleRuntime is passed to Style plugin which needed this changes for the functionalities to work. We will merge this two into one.

File: src/LinkSetURLCommand.js [hunk 4]
```diff
@@ -43,15 +43,42 @@ class LinkSetURLCommand extends UICommand {
...
-  waitForUserInput = (
+  showTocList = async (view) => {
+    let storeTOCvalue = [];
+    const TOCselectedNode = [];
+
+    const stylePromise = view.runtime;
+
+    if (stylePromise === null || undefined) {
+      return TOCselectedNode;
+    } else {
```
Comment (Tezra) [inline src/LinkSetURLCommand.js]: This else block is unnecessary as it is just the rest of the function. You can remove it to reduce nesting.

File: src/LinkSetURLCommand.js [hunk 5]
```diff
@@ -43,15 +43,42 @@ class LinkSetURLCommand extends UICommand {
...
+
+      const prototype = Object.getPrototypeOf(stylePromise);
+      const styles = await prototype.getStylesAsync();
+
+      storeTOCvalue = styles
+        .filter((style) => style.styles.toc === true)
+        .map((style) => style.styleName);
+
+      view.state.tr.doc.descendants((node, pos) => {
+        if (node.attrs.styleName) {
```
Comment (Tezra) [inline src/LinkSetURLCommand.js]: These 3 blocks can be simplified to `if(storeTocValue.contains(node.attrs.stylename))`.

File: src/LinkSetURLCommand.js [hunk 6]
```diff
@@ -43,15 +43,42 @@ class LinkSetURLCommand extends UICommand {
     return from < to;
   };
 
-  waitForUserInput = (
+  showTocList = async (view) => {
```
Comment (Tezra) [inline src/LinkSetURLCommand.js:48]: What exactly is this function for? Licit does not need to know about the TOC, and what does this have to do with InnerLink?

File: src/LinkTooltipPlugin.js [hunk 7]
```diff
@@ -41,6 +41,18 @@ class LinkTooltipView {
     this.update(editorView, null);
   }
 
+  getInnerlinkSelected_position(view: EditorView, selectionId): void {
+    let tocItemPos;
+    view.state.tr.doc.descendants((node, pos) => {
+      if (node.attrs.styleName && node.attrs.innerLink) {
```
Comment (Tezra) [inline src/LinkTooltipPlugin.js]: These 2 ifs can be combined.

File: .eslintrc.cjs [hunk 8]
```diff
@@ -10,7 +10,7 @@ module.exports = {
   },
   plugins: ['react'],
   rules: {
-    'react/jsx-sort-props': 'error',
```
Comment (Tezra) [inline .eslintrc.cjs]: No disabling lint rules

File: src/ui/LinkURLEditor.js [hunk 9]
```diff
@@ -61,33 +89,125 @@ class LinkURLEditor extends React.PureComponent<any, any> {
...
+  handleOptionChange = (
+    textContent_,
+    tocNodePosition_,
+    view: EditorView
+  ) => {
+    const innerString = 'INNER______LINK';
+    const tr = view.state.tr;
+    const TocNode = view.state.doc.nodeAt(tocNodePosition_);
+    let textContent;
+    if (TocNode && (TocNode.attrs.innerLink == null || TocNode.attrs.innerLink == '' || TocNode.attrs.innerLink == undefined)) {
```
Comment (Tezra) [inline src/ui/LinkURLEditor.js]: can be simplified to `if(TocNode?.attrs?.innerLink)`

File: src/ui/LinkURLEditor.js [hunk 10]
```diff
@@ -61,33 +89,125 @@ class LinkURLEditor extends React.PureComponent<any, any> {
...
       </div>
     );
   }
 
+  handleOptionChange = (
+    textContent_,
+    tocNodePosition_,
+    view: EditorView
+  ) => {
+    const innerString = 'INNER______LINK';
```
Comment (Tezra) [inline src/ui/LinkURLEditor.js]: All instances of these should all be referencing the same const to insure they are all using the same text.

File: src/ui/RichTextEditor.js [hunk 11]
```diff
@@ -88,6 +89,7 @@ class RichTextEditor extends React.PureComponent<any, any> {
           placeholder={placeholder}
           readOnly={readOnly}
           runtime={runtime}
+          styleRuntime={styleRuntime}
```
Comment (Tezra) [inline src/ui/RichTextEditor.js]: What does the styleRuntime have to do with InnerLink?

**Review by Tezra — CHANGES_REQUESTED** [review]

File: src/ui/LinkURLEditor.js (hunk 9 above)
Comment (Jayakrishnan-MS) [reply src/ui/LinkURLEditor.js]: I guess 'if(TocNode?.attrs?.innerLink)'  if we give like this -the condition will only pass if innerLink has value right?. But we want it to enter the 'if' condition when innerLink is null, empty string or undefined.

Comment (Tezra) [reply src/ui/LinkURLEditor.js]: false, 0, '', null, and undefined are all falsy values, so can all be treated as "false". so `if(!TocNode?.attrs?.innerLink)`

Comment (Jayakrishnan-MS) [reply src/ui/LinkURLEditor.js]: Alright

File: src/LinkSetURLCommand.js (hunk 6 above)
Comment (rohanrjohn) [reply src/LinkSetURLCommand.js:48]: For innerlink functionality, we are displaying toc selected style para list to choose from and create a link between them (jump functionality). For this we are using these style functionalities.

File: src/ui/RichTextEditor.js (hunk 11 above)
Comment (rohanrjohn) [reply src/ui/RichTextEditor.js]: Have removed it, as styleRuntime is no longer needed.

File: src/Types.js [hunk 12]
```diff
@@ -1,6 +1,7 @@
 // @flow
 
 import * as React from 'react';
+import { Style } from '@modusoperandi/licit-custom-styles';
```
Comment (Tezra) [inline src/Types.js]: If you're importing from custom styles, why not just import the runtime? Also, I don't see this in the peer dependency list.

**Review by Tezra — CHANGES_REQUESTED** [review]

File: licit/client/CustomLicitRuntime.js [hunk 13]
```diff
@@ -14,13 +18,26 @@ import { POST } from '../../src/client/http.js';
...
 // const GLOSSARY_URI = 'http://greathints.com:3003';
 
+type Style = {
+  /**
+   * Name of the style. Case insensitive value must be unique.
+   */
+  styleName: string; // Style name to display
+  mode?: number; // For Style Editor UI behaviour //0 = new , 1- modify, 2- rename, 3- editall
+  description?: string; // style description
+  styles?: HTMLStyles;
```
Comment (Tezra) [inline licit/client/CustomLicitRuntime.js:57]: lint issue. HTMLStyles is not defined.

File: src/Types.js (hunk 12 above)
Comment (rohanrjohn) [reply src/Types.js]: Have added Style type to avoid custom styles dependency

File: licit/client/CustomLicitRuntime.js (hunk 13 above)
Comment (rohanrjohn) [reply licit/client/CustomLicitRuntime.js:57]: Have added the HTMLStyles type.

**Review by Tezra — APPROVED** [review]

**Review by Tezra — CHANGES_REQUESTED** [review]
Needs resync

**Comment by Tezra** [comment]
Issue with Jest preventing tests from running.

File: licit/client/CustomLicitRuntime.js [hunk 14]
```diff
@@ -3,7 +3,11 @@
 // This implements the interface of `EditorRuntime`.
 // To  run  editor directly:
 import type { ImageLike, RecentColor } from '../../src/Types.js';
-import { POST } from '../../src/client/http.js';
+import { POST, GET, DELETE, PATCH } from '../../src/client/http.js';
+
+
+const STYLES_URI = 'http://greathints.com:3000';
```
Comment (Tezra) [inline licit/client/CustomLicitRuntime.js:9]: This URL isn't being used, and shouldn't be hardcoded.

File: src/client/Licit.js [hunk 15]
```diff
@@ -630,6 +634,27 @@ class Licit extends React.Component<any, any> {
     });
   };
 
+  exportJSON = () => {
```
Comment (Tezra) [inline src/client/Licit.js]: This function in unnecessary. To call it, the caller will already have access to the doc and the export plugin. For type safety, the export function should be accessed directly.

**Review by Tezra — CHANGES_REQUESTED** [review]

File: licit/client/CustomLicitRuntime.js (hunk 14 above)
Comment (rohanrjohn) [reply licit/client/CustomLicitRuntime.js:9]: This is for internal test purpose used in the working of standalone editor

**Review by Tezra — APPROVED** [review]

**Review by rodgersgb — APPROVED** [review]
